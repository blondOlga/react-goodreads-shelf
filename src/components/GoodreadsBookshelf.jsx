import React from "react";
import Xml2JsUtils from "../util/xml2js-utils";
import Book from "./Book";
import Loader from "./Loader";

const shelfStyle = (minWidth) => ({
	display: "grid",
	alignItems: "center",
	gridTemplateColumns: `repeat(auto-fit, minmax(${minWidth}px, 1fr))`,
	gridColumnGap: "1vw",
	margin: "1vw",
	textAlign: "center",
});

class GoodreadsBookshelf extends React.Component {

	constructor(props) {
		super(props);
		this.state = { 
			books: [], 
			loaded: false,
			options: props.options || { width: 100 }
		};
	}

	componentWillMount() {
		this.getBooks();
	}

	getUrl() {
		// Build a request to the Goodreads API
		const url = new URL(`https://cors-anywhere.herokuapp.com/https://www.goodreads.com/review/list/${this.props.userId}`);
		url.searchParams.set("key", this.props.apiKey);
		url.searchParams.set("per_page", this.props.limit || 10);
		url.searchParams.set("shelf", this.props.shelf || "read");
		url.searchParams.set("sort", this.props.sort || "date_read");
		url.searchParams.set("v", 2);
		return url;
	}

	async getBooksJson() {
		if (typeof window !== "undefined" && window.fetch) {
			const url = this.getUrl();
			const response = await fetch(url);
			const xmlText = await response.text();

			const json = Xml2JsUtils.parse(xmlText);
			return json.GoodreadsResponse.reviews.review; // This is where the list of books is stored
		} else {
			throw "Error: fetch is not defined in this environment";
		}
	}

	async getBooks() {

		try {
			const books = await this.getBooksJson();
	
			this.setState({
				books: books,
				loaded: true
			});
	
		} catch (e) {

			console.error(e);

			// Indicate that we errored
			this.setState({
				loaded: true,
				error: true
			});
		}
		
		
	}

	render() {
		return (
			<div class="goodreads-shelf">
				<div style={shelfStyle(this.state.options.width)}>
					{this.state.books.map(book => {
						return <Book key={book.id} book={book.book} options={this.state.options} />
					})}
				</div>
				{ this.state.loaded ? "" : <Loader /> }
				{ this.state.error ? <div>Sorry, we couldn't load books right now</div> : "" }
			</div>
			
			
		)
	}

}

export default GoodreadsBookshelf;