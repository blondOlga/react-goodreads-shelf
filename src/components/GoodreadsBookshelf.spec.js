import GoodreadsBookshelf from "./GoodreadsBookshelf";
import React from "react";
import { shallow, mount, render } from "enzyme";

describe("testing bookshelf", () => {

	beforeEach(() => {
		fetch.resetMocks();
		fetch.mockResponseOnce(JSON.stringify({ GoodreadsResponse: { reviews: { review: [] } } }));
	})

	it("renders without crashing", () => {
		const shelf = shallow(<GoodreadsBookshelf />);
		expect(shelf).toMatchSnapshot();
	});
	
	it("passes props properly", () => {
		const shelf = shallow(<GoodreadsBookshelf apiKey="test" userId="testUser" limit={15} shelf="read" sort="date_read" />);
		const props = shelf.instance().props;
		expect(props.limit).toEqual(15);
		expect(props.userId).toEqual("testUser");
		expect(props.apiKey).toEqual("test");
		expect(props.shelf).toEqual("read");
		expect(props.sort).toEqual("date_read");
	});
	
	// it("gets books", async () => {
	// 	const shelf = shallow(<GoodreadsBookshelf apiKey="test" userId="testUser" limit={15} shelf="read" sort="date_read" />);
	// 	const books = await shelf.instance().getBooksJson();
	// 	await expect(books.length).resolves.toBeLessThanOrEqual(15);
	// });
});