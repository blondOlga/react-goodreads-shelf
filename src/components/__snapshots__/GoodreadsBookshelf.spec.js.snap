// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `ShallowWrapper {}`;

exports[`testing bookshelf renders without crashing 1`] = `ShallowWrapper {}`;
